### PR TITLE
Implement AWS S3 backend

### LINKED ISSUE #34: AWS S3 support
This is a proposal to support S3 as a backend in this crate. I'm working on a first implementation of such a backend and was wondering if this is something you'd agree to have and when ready release with self-update crate?

I'm doing this as part of task for some binaries where we use this crate from (I'm working on https://github.com/maidsafe project), and we publish some releases on S3, therefore I was planning to send it as a PR in the next few days if that's something you'll consider for this crate.

### PR DESCRIPTION
**Pull request by bochaco**
Resolves #34

### CODE REVIEW

File: src/backends/s3.rs [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+            _ => (), // There are several other `Event`s we ignore here
+        }
+
+        buf.clear();
+    }
+
+    Ok(releases)
+}
+
+fn add_release(releases: &mut Vec<Release>, mut rel: Release) {
```
Comment (jaemk) [inline src/backends/s3.rs]: Could you mention this is handling deduping in a doc string?

**Comment by bochaco** [comment]
Thanks @jaemk , I'll work on your comments tomorrow sending a new commit to fix them.

File: src/update.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+use crate::{confirm, errors::*, version, Download, Extract, Move, Status};
+
+/// Release asset information
+#[derive(Clone, Debug, Default)]
+pub struct ReleaseAsset {
+    pub download_url: String,
+    pub name: String,
+}
+
+/// Update status with extended information from Github
```
[the remark this answers is not in the document]
Comment (bochaco) [reply src/update.rs]: No, it's common to any backend, I'll change that doc.

File: src/lib.rs [hunk 3]
```diff
@@ -42,6 +42,30 @@ fn update() -> Result<(), Box<::std::error::Error>> {
 
 Run the above example to see `self_update` in action: `cargo run --example github`
 
+Amazon S3 is also supported as the backend to check for new releases. Provided a `bucket_name`
+and `asset_prefix` string, `self_update` will look up all matching files using the following format
+as a convention for the filenames: `<asset name>-<semver>-<platform/target>.<extension>`.
```
Comment (jaemk) [inline src/lib.rs]: is `asset name` here the meant to be `asset_prefix`?

File: src/lib.rs [hunk 4]
```diff
@@ -42,6 +42,30 @@ fn update() -> Result<(), Box<::std::error::Error>> {
...
+Any file not matching the format, or not matching the provided prefix string, is be ignored.
```
Comment (jaemk) [inline src/lib.rs]: small nit: "will be"

File: src/lib.rs (hunk 3 above)
Comment (bochaco) [reply src/lib.rs]: No, the asset name is expected to be the name of the asset that is being published as released, followed by the `-` with semver , etc., that's how we parse it for figuring out the name and version of an asset.
The asset prefix is something to filter them out when fetching them from S3, it's matched against the filename, and it's used as part of the API URL itself: `https://<bucket name>.s3.amazonaws.com/?list-type=2&max-keys=<N>&prefix=<asset prefix>`

**Comment by bochaco** [comment]
Thanks a lot @jaemk! Do you think a new release can be published so we can start using it from one of our crates?

**Comment by jaemk** [comment]
Yes! I ran out of time yesterday, but I'll get to it today

**Comment by bochaco** [comment]
oh nice! thanks a lot @jaemk !

**Pull request merged**